### ISSUE TITLE
No SSL for JMX/JNDI connection

### ISSUE THREAD

**Issue by mmatela** [oracle/visualvm#267]
I have a server that can be connected to with a URL like `service:jmx:rmi:///jndi/rmi://127.0.0.1:1099/jmxrmi` that uses an `SslRMIServerSocketFactory`.
When I use this URL in the "Add JMX Connaction" dialog in VisualVM it shows a message: "The connection to admin@service... could not be made using SSL. Retry without SSL?" Looking at the sent network packets I see that no SSL connection attempt was made.

This is similar to https://bugs.java.com/bugdatabase/view_bug.do?bug_id=8020207 for jconsole. There they suggest to use VisualVM as a workaround, but it turns out it doesn't work either.

I did some debugging and the problem seems to be in https://github.com/oracle/visualvm/blob/0588a4cc9af99d1584f0be4a682be40d9d14386e/visualvm/jmx/src/org/graalvm/visualvm/jmx/impl/ProxyClient.java#L285
It tries to catch an `IOException` and then setup an SSL socket with `env.put("com.sun.jndi.rmi.factory.socket", sslRMIClientSocketFactory);`. The actual exception thrown is not `IOException`, but `SecurityException`, thrown by `javax.management.remote.rmi.RMIConnector.checkStub(Remote, Class<?>)`. So it should be enough to catch both exceptions in line 286. Or maybe the SSL socket should be setup earlier, along with `env.put("jmx.remote.x.check.stub", "true");` in line 277.

**Comment by thurka** [comment]
Is there a way to reproduce this problem? Can you provide step-by-step instructions? How is your case different from standard JMX setup with SSL?

**Comment by mmatela** [comment]
After further investigation I realized the problem was in my server setup. I didn't know about `jmx.remote.rmi.client.socket.factory` and `jmx.remote.rmi.server.socket.factory`, after setting proper SSL factories there VisualVM could connect correctly. (And the fact that my server tried to be smart and properly handle unsecured connections as well made my debug session even more confusing).

I also found out that jconsole can connect as well, with the simpler `127.0.0.1:1099` notation instead of of the full URL (but only if the actual service url ends with `/jmxrmi`.

Sorry about the noise!
